node/network/collator-protocol/src/collator_side.rs

```diff
@@ -91,6 +90,96 @@ impl metrics::Metrics for Metrics {
 	}
 }
 
+/// The group of validators that is assigned to our para at a given point of time.
+///
+/// This structure is responsible for keeping track of which validators belong to a certain group for a para. It also
+/// stores a mapping from [`PeerId`] to [`ValidatorId`] as we learn about it over the lifetime of this object. Besides
+/// that it also keeps track to which validators we advertised our collation.
+struct ValidatorGroup {
+	/// All [`ValidatorId`]'s that are assigned to us in this group.
+	validator_ids: HashSet<ValidatorId>,
+	/// The mapping from [`PeerId`] to [`ValidatorId`]. This is filled over time as we learn the [`PeerId`]'s from the
+	/// authority discovery. It is not ensured that this will contain *all* validators of this group.
+	peer_ids: HashMap<PeerId, ValidatorId>,
+	/// All [`ValidatorId`]'s of the current group to that we advertised our collation.
+	advertised_to: HashSet<ValidatorId>,
+}
+
+impl ValidatorGroup {
+	/// Returns `true` if we should advertise our collation to the given peer.
+	fn should_advertise_to(&self, peer: &PeerId) -> bool {
+		match self.peer_ids.get(peer) {
+			Some(validator_id) => !self.advertised_to.contains(validator_id),
+			None => false,
+		}
+	}
+
+	/// Should be called after we advertised our collation to the given `peer` to keep track of it.
+	fn advertised_to_peer(&mut self, peer: &PeerId) {
+		if let Some(validator_id) = self.peer_ids.get(peer) {
+			self.advertised_to.insert(validator_id.clone());
+		}
+	}
+
+	/// Add a [`PeerId`] that belongs to the given [`ValidatorId`].
+	///
+	/// This returns `true` if the given validator belongs to this group and we could insert its [`PeerId`].
+	fn add_peer_id_for_validator(&mut self, peer_id: &PeerId, validator_id: &ValidatorId) -> bool {
+		if !self.validator_ids.contains(validator_id) {
+			false
+		} else {
+			self.peer_ids.insert(peer_id.clone(), validator_id.clone());
+			true
+		}
+	}
+}
+
+impl From<HashSet<ValidatorId>> for ValidatorGroup {
+	fn from(validator_ids: HashSet<ValidatorId>) -> Self {
+		Self {
+			validator_ids,
+			peer_ids: HashMap::new(),
+			advertised_to: HashSet::new(),
+		}
+	}
+}
+
+/// Wrapper around the connection requests per relay parent.
+///
+/// Implements [`Future`] to check if any of the internal requests returned a result to propagate it to the calling
+/// context. It is safe to call this future multiple times, if there are no requests or none of them returned a result
+/// (because they returned either [`Poll::Pending`]/`None`) it just returns [`Poll::Pending`].
+#[derive(Default)]
+struct ConnectionRequests {
+	requests: HashMap<Hash, validator_discovery::ConnectionRequest>,
+}
+
```

Comment:
I've ran into the same thing and have created #1966 

What do you think @ordian ?

Reply:
I'm not sure this is the same thing. In this implementation it's possible to revoke some requests based on the relay_parent, whereis in @montekki's PR we have one large requests that can only be cancelled at once.

In either case, this logic should go in `node/subsystem-util/src/validator_discovery.rs` for reusability and better separation of concerns.

Reply:
will two concurrent `ConnectToValidators` work correctly?

Reply:
If they will then I think `ConnectionRequests` is a more handy API; just let's make it reusable

Reply:
> will two concurrent `ConnectToValidators` work correctly?

Yes, I don't see a reason why not.